### ISSUE TITLE
[BUG][Fuzzer][wrong-code] `T.cumsum`/`T.cummax` over a non-contiguous 2-D shared sub-region silently returns wrong numbers instead of a correct prefix scan

### ISSUE THREAD

**Issue by Hughshine** [tile-ai/tilelang#2523]
### Required prerequisites

- [x] I have read the documentation <https://tilelang.com>.
- [x] I have searched the [Issue Tracker](https://github.com/tile-ai/tilelang/issues) that this hasn't already been reported. (comment there if it has.)

### What version of TileLang are you using?

0.1.11

### System information

- Installed from the `0.1.11` wheel; source pinned to tag `v0.1.11` (commit `cd37ed5fc35ae7a60a1277c8eb49028174ac51e6`).
- Python 3.12.3, linux
- GPU: **NVIDIA H100 80GB HBM3** (compute capability 9.0, sm_90). The defect is in the architecture-independent shared-scan lowering and CUDA device template, so it is not Hopper-specific; it was reproduced on sm_90.

```python
import sys, tilelang, torch
print(sys.version.split()[0], sys.platform)        # 3.12.3 linux
print(tilelang.__version__)                         # 0.1.11
print(torch.__version__, torch.version.cuda)        # 2.12.1+cu130 13.0
print(torch.cuda.get_device_name(), torch.cuda.get_device_capability())  # NVIDIA H100 80GB HBM3 (9, 0)
```

### Problem description

**Silent miscompile:** `T.cumsum` over a 2-D **slice** of a shared buffer (`T.cumsum(src=big[0:M, 0:N], …)` where `big` is allocated `(M, NBIG)` with `NBIG > N`) compiles and runs with no error but returns **wrong numbers**, deterministically. The scan lowering passes the *slice* extents `H=M, W=N` to the device kernel with **no row-pitch argument**, and the template uses `W` as the row stride (`src + row*W`). But the buffer's true row pitch is `NBIG`, not `N` — so row 0 is correct and every subsequent row is mis-addressed by `(NBIG − N)` elements, corrupting the scan. The generated call literally reads `CumSum2D<...>::run(&big[0], &big[0], 8, 40)` while `big` has pitch 64. When the region is **contiguous** (`NBIG == N`) the assumption holds and the result is bit-exact for both dims (control below), so the defect is exactly the row-pitch ≠ scan-width handling — not the scan arithmetic. The same `InclusiveScan2D` template backs `CumMax2D`, so `T.cummax` has the identical defect.

<details>
<summary>Trigger-boundary table (all rows run fresh on the H100, float32, exact-int inputs)</summary>

| `M` | `N` | `NBIG` | `dim` | `reverse` | result |
|-----|-----|--------|-------|-----------|--------|
| 8  | 40 | 40 | 0 | no  | **equal** (contiguous control) |
| 8  | 40 | 40 | 1 | no  | **equal** (contiguous control) |
| 8  | 40 | 64 | 0 | no  | **wrong** (`equal=False`, deterministic) |
| 8  | 40 | 64 | 1 | no  | **200/320 wrong** |
| 40 | 8  | 64 | 0 | no  | **258/320 wrong** |
| 33 | 8  | 48 | 0 | no  | **197/264 wrong** |
| 8  | 40 | 64 | 1 | yes | **196/320 wrong** |
| 40 | 8  | 64 | 0 | yes | **284/320 wrong** |

The contiguous (`NBIG == N`) controls are bit-exact for both dims, isolating the defect to the row-pitch ≠ scan-width handling of the 2-D scan — not to `T.cumsum`'s scan arithmetic, not to the `T.Parallel` load/store, and not to the harness.
</details>

<details>
<summary>Root cause — lowering drops the region stride; device template uses <code>W</code> as the row pitch</summary>

**(1) The lowering passes slice extents and no stride.** `LowerSharedScan` collects the per-axis **region** extents ([`scan.h#L45-L48`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/backend/common/op/scan.h#L45-L48)) and for the 2-D case builds `args = {sym, src_ptr, dst_ptr, src_extents[0], src_extents[1]}` ([`#L57-L61`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/backend/common/op/scan.h#L57-L61)) — `H, W` = the slice's logical shape, **no** row-pitch argument. The base pointer `MakeAccessPtrFromRegion` ([`#L42`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/backend/common/op/scan.h#L42)) is correctly offset via the buffer's true strides ([`utils.cc#L95-L112`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/op/utils.cc#L95-L112)), but only the flat base + slice extent are conveyed; the stride is dropped.

**(2) The device template uses `W` as the row pitch.** `InclusiveScan2D::run(src, dst, H, W)` ([`scan.h#L118-L145`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/scan.h#L118-L145)) treats `W` as the leading-dim stride: `Axis==1` scans `src + row*W` ([`#L133-L134`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/scan.h#L133-L134)); `Axis==0` scans columns with element stride `W` ([`#L142`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/scan.h#L142)). So it **assumes the sub-region is contiguous** (pitch == `W`); for `NBIG > W` all rows after the first are mis-addressed.

Op path: `CumSumOpNode::Lower` → `LowerSharedScan` ([`scan.cc#L139-L142`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/op/scan.cc#L139-L142)); the same `InclusiveScan2D` backs `CumMax2D` ([`scan.h#L175-L182`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/scan.h#L175-L182)).
</details>

### Reproducible example code

```python
import tilelang as tl
import tilelang.language as T
import torch
import re


def make_kernel(M, N, NBIG, dim, reverse, threads):
    @T.prim_func
    def cumsum(
        A: T.Tensor((M, N), "float32"),
        B: T.Tensor((M, N), "float32"),
    ):
        with T.Kernel(1, threads=threads) as bx:
            big = T.alloc_shared((M, NBIG), "float32")     # row pitch = NBIG
            for i, j in T.Parallel(M, NBIG):
                big[i, j] = 0.0
            for i, j in T.Parallel(M, N):
                big[i, j] = A[i, j]
            # scan the leading N columns of a width-NBIG buffer -> non-contiguous when NBIG > N
            T.cumsum(src=big[0:M, 0:N], dst=big[0:M, 0:N], dim=dim, reverse=reverse)
            for i, j in T.Parallel(M, N):
                B[i, j] = big[i, j]
    return cumsum


def ref(A, dim, reverse):
    if reverse:
        return A.flip(dims=[dim]).cumsum(dim=dim).flip(dims=[dim])
    return A.cumsum(dim=dim)


def run_once(M, N, NBIG, dim, reverse=False, threads=256, seed=3):
    k = tl.compile(make_kernel(M, N, NBIG, dim, reverse, threads), out_idx=-1)
    torch.manual_seed(seed)
    A = torch.randint(-2, 3, (M, N), dtype=torch.float32).cuda()  # exact-int oracle
    out = k(A)
    exp = ref(A, dim, reverse)
    return torch.equal(out, exp), out, exp, k


print("--- contiguous baseline (NBIG==N) should be CORRECT ---")
for dim in (0, 1):
    eq, *_ = run_once(8, 40, 40, dim)
    print(f"  NBIG=N=40 dim={dim}: equal={eq}")     # -> equal=True (both dims)

print("--- strided region (NBIG>N) WRONG + determinism ---")
eq1, out1, _, k = run_once(8, 40, 64, 0)
eq2, out2, _, _ = run_once(8, 40, 64, 0)
print(f"  NBIG=64 N=40 dim=0: run1 equal={eq1}, run2 equal={eq2}, det(run1==run2)={torch.equal(out1, out2)}")
# -> run1 equal=False, run2 equal=False, det(run1==run2)=True
src = k.get_kernel_source()
print("  codegen call:", re.findall(r"CumSum2D<[^>]*>::run\([^;]*", src)[0][:120])
# -> CumSum2D<256, 0, false>::run((&(big[0])), (&(big[0])), 8, 40)
print("  -> passes W=40 (slice extent) but buffer row stride is 64 -> mis-addressed")

print("--- more shapes / dims / reverse ---")
for (M, N, NBIG, dim, rev) in [(8, 40, 64, 1, False), (40, 8, 64, 0, False),
                               (33, 8, 48, 0, False), (8, 40, 64, 1, True),
                               (40, 8, 64, 0, True)]:
    eq, out, exp, _ = run_once(M, N, NBIG, dim, rev)
    nmis = int((out != exp).sum().item())
    print(f"  M={M} N={N} NBIG={NBIG} dim={dim} rev={rev}: equal={eq} mis={nmis}/{M*N}")
# -> M=8  N=40 NBIG=64 dim=1 rev=False: equal=False mis=200/320
# -> M=40 N=8  NBIG=64 dim=0 rev=False: equal=False mis=258/320
# -> M=33 N=8  NBIG=48 dim=0 rev=False: equal=False mis=197/264
# -> M=8  N=40 NBIG=64 dim=1 rev=True:  equal=False mis=196/320
# -> M=40 N=8  NBIG=64 dim=0 rev=True:  equal=False mis=284/320
```

The output (annotated inline above with `# ->`) is stable across repeated runs on the H100; the contiguous (`NBIG == N`) controls are bit-exact for both dims, isolating the defect to the row-pitch ≠ scan-width handling.

### Traceback

```pytb
No traceback — the kernel compiles and runs to completion; the result is silently wrong and deterministic.
```

### Expected behavior

The core defect is that the lowering **accepts a sliced 2-D region, compiles it with no error, and then silently returns wrong numbers**. That is a bug under either reading of the intended contract:

- **If sliced regions are meant to be supported** — which the docstring's wording suggests: it says `T.cumsum` *"Supports Buffer, BufferRegion, and BufferLoad inputs, allowing operations on buffer slices/regions"* (though its only slice example is 1-D, `T.cumsum(InputG_fragment[i*chunk:(i+1)*chunk], dim=0)`, which is inherently contiguous and so cannot exhibit this) — then the 2-D non-contiguous slice must produce the correct prefix scan (`torch.equal` against `torch.cumsum`), matching the contiguous case.
- **If a 2-D region whose width differs from the buffer's row pitch is not meant to be supported** — then the lowering must **reject it at compile time** with a clear diagnostic.

Either resolution is fine and is the maintainer's call; the current behavior is neither — it silently miscomputes. The contiguous case (`NBIG == N`) of the identical program is bit-exact, confirming the scan arithmetic itself is correct and only the region-pitch handling is at fault.

### Additional context

**Suggested fix (proposed, not built or tested).** Thread the buffer's true row pitch (leading-dimension stride, in elements) through to the scan kernel and use it for row/column addressing instead of the slice width `W`.

1. **Lowering** ([`src/backend/common/op/scan.h#L57-L61`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/backend/common/op/scan.h#L57-L61), 2-D branch): in addition to `src_extents[0]`/`[1]`, compute the buffer's leading-dimension stride for the scanned region (for a row-major 2-D shared buffer this is the buffer's full row width `NBIG`, i.e. `op.src->shape[op.src->shape.size()-1]`, not the region extent) and pass it as an extra `run()` argument, e.g. `args = {sym, src_ptr, dst_ptr, src_extents[0], src_extents[1], stride}`. Both src and dst regions must share the same pitch, so `ICHECK` they are equal (the template uses one stride for both).
2. **Device template** ([`src/tl_templates/cuda/scan.h#L118-L145`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/scan.h#L118-L145)): extend `CumSum2D::run` / `InclusiveScan2D::run` (and `CumMax2D`) to take an extra `int stride` (row pitch). Replace the `W`-based row addressing: `Axis == 1` → `src + row * stride` / `dst + row * stride` ([`#L133-L134`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/scan.h#L133-L134)); `Axis == 0` → `InclusiveScanLine(src + col, dst + col, H, /*stride=*/stride)` ([`#L142`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/scan.h#L142)). Keep `W` as the column count / extent bound and `stride` as the row pitch; when the region is contiguous (`stride == W`) behavior is unchanged, so the passing controls stay green. This is symmetric for `CumMax2D`, which shares `InclusiveScan2D`.

Both directions are low-cost, so "at least don't miscompile silently" is a cheap bar to clear:

- **Reject (≈5 lines, front-end only):** add an `ICHECK` in `LowerSharedScan`'s 2-D branch that the region's last-dim extent equals the backing buffer's last-dim shape (`src_extents[1] == op.srcRegion_->buffer->shape.back()`) — `LowerSharedScan` already has both values in hand. Turns the silent miscompile into a hard compile error.
- **Support (≈4 mechanical edits):** the threading fix above. Notably the innermost `InclusiveScanLine` is *already* `stride`-based (`src[(base+lane)*stride]`, [`scan.h#L46`](https://github.com/tile-ai/tilelang/blob/cd37ed5fc35ae7a60a1277c8eb49028174ac51e6/src/tl_templates/cuda/scan.h#L46)), so supporting a real pitch is just plumbing one extra argument down through `CumSum2D`/`CumMax2D`/`InclusiveScan2D`; when `stride == W` behavior is byte-identical, so the passing controls stay green.

I have not built or verified any of these changes.

**Provenance.** Both the scan lowering (`src/backend/common/op/scan.h`) and the device template (`src/tl_templates/cuda/scan.h`) were introduced together in [#2262 "[TileOP] Add scan operators"](https://github.com/tile-ai/tilelang/pull/2262) (commit `146a1d3b`, merged 2026-05-25); the slice-extent-only `args` line has been present since that commit (via `git log -S'src_extents[1]'`), i.e. the defect has shipped since scan operators were first added. Present in v0.1.11.

I searched the open and closed `tile-ai/tilelang` tracker and found no existing report of this defect (the 2-D scan lowering passing the slice extent as the device kernel's row stride for a non-contiguous region). The nearest is **#2284** (closed) *"`T.cumsum` dim=0 illegal memory access on rectangular 2D shared tiles"* — also `T.cumsum` 2-D, but the failure there is a **crash** (`CUDA_ERROR_ILLEGAL_ADDRESS`); here the buffer addressing stays in bounds (the region is a slice of a strictly larger allocation) and the failure is a **silent wrong result**, not a crash.
